src/docregistry/registry.ts

```diff
@@ -717,24 +508,438 @@ class DocumentRegistry implements IDocumentRegistry {
+     */
+    readonly defaultKernelLanguage: string;
+
+    /**
+     * Serialize the model to a string.
+     */
+    toString(): string;
+
+    /**
+     * Deserialize the model from a string.
+     *
+     * #### Notes
+     * Should emit a [contentChanged] signal.
+     */
+    fromString(value: string): void;
+
+    /**
+     * Serialize the model to JSON.
+     */
+    toJSON(): any;
+
+    /**
+     * Deserialize the model from JSON.
+     *
+     * #### Notes
+     * Should emit a [contentChanged] signal.
+     */
+    fromJSON(value: any): void;
+  }
+
+  /**
+   * The document context object.
+   */
+  export
+  interface IContext<T extends IModel> extends IDisposable {
+    /**
+     * A signal emitted when the kernel changes.
+     */
+    kernelChanged: ISignal<this, IKernel>;
+
+    /**
+     * A signal emitted when the path changes.
+     */
+    pathChanged: ISignal<this, string>;
+
+    /**
+     * A signal emitted when the contentsModel changes.
+     */
+    fileChanged: ISignal<this, Contents.IModel>;
+
+    /**
+     * A signal emitted when the context is fully populated for the first time.
+     */
+    populated: ISignal<this, void>;
+
+    /**
+     * A signal emitted when the context is disposed.
+     */
+    disposed: ISignal<this, void>;
+
+    /**
+     * Get the model associated with the document.
+     */
+    readonly model: T;
+
+    /**
+     * The current kernel associated with the document.
+     */
+    readonly kernel: IKernel;
+
+    /**
+     * The current path associated with the document.
+     */
+    readonly path: string;
+
+    /**
+     * The current contents model associated with the document
+     *
+     * #### Notes
+     * The model will have an empty `contents` field.
+     * It will be `null` until the context is populated.
+     */
+    readonly contentsModel: Contents.IModel;
+
+    /**
+     * Get the kernel spec information.
+     */
+    readonly kernelspecs: Kernel.ISpecModels;
+
+    /**
+     * Test whether the context is fully populated.
+     */
+    readonly isPopulated: boolean;
+
+    /**
+     * Change the current kernel associated with the document.
+     *
+     * #### Notes
+     * If no options are given, the session is shut down.
+     */
+    changeKernel(options?: Kernel.IModel): Promise<IKernel>;
+
+    /**
+     * Save the document contents to disk.
+     */
+    save(): Promise<void>;
+
+    /**
+     * Save the document to a different path chosen by the user.
+     */
+    saveAs(): Promise<void>;
+
+    /**
+     * Revert the document contents to disk contents.
+     */
+    revert(): Promise<void>;
+
+    /**
+     * Create a checkpoint for the file.
+     *
+     * @returns A promise which resolves with the new checkpoint model when the
+     *   checkpoint is created.
+     */
+    createCheckpoint(): Promise<Contents.ICheckpointModel>;
+
+    /**
+     * Delete a checkpoint for the file.
+     *
+     * @param checkpointID - The id of the checkpoint to delete.
```

Comment:
Could we change instances of `checkpointID` to read `checkpointId` instead, with the lowercase `d`? It's more consistent the rest of our code.